### DISCUSSION TITLE
0.51.5 question re osxphotos.PhotosDB(...)

### DISCUSSION THREAD

**Discussion by WadeJA** [RhetTbull/osxphotos#781] [General]
First off, thank you so much for providing such a comprehensive library; I'm looking forward to using it to export a list of what files contain a certain face. Meaning I had set up a rather large directory (w/subdirectories) of JPGs which I then imported into Photos. Once the facial recognition process had completed, I have now been ready to determine which file(s) (& in which original Dirs) contain a given face. Hence the desire to use your library. Given this context, now to my question:

In many of your tests & examples, you use something like
photosdb = osxphotos.PhotosDB(db)
But I do not see PhotosDB, either in the library or in the documentation. Perhaps I didn't look very well, but has this changed to something different? Or perhaps I am (obviously?) missing something. Any helps / hints would be greatly appreciated.

**Comment by WadeJA** [comment]
Sorry, I should have listed info in the prior msg: Using a new Mac Air M2 with OS 12.4 and Photos 7.0 (build 450.0.160), if that helps, in Python 3.10.6

**Reply by RhetTbull** [reply]
Hi @WadeJA Take a look [here](https://rhettbull.github.io/osxphotos/reference.html?highlight=photosdb#osxphotos.PhotosDB).  Also, take a look at the [API README](https://github.com/RhetTbull/osxphotos/blob/master/API_README.md) and then the [example](https://github.com/RhetTbull/osxphotos/tree/master/examples) code.  

So that I can make the documentation more clear, where where you looking that you didn't see PhotosDB?  Not too many people use the library to write their own code (though that's the reason I originally built this) so I spend less time on documenting the API  vs documenting the export command which is what most people want.

**Comment by WadeJA** [comment]
In the example code, line 11, I see that it is referenced. Also, in the API README, I certainly see the reference; same for the documentation. But (sorry I wasn't clear enough), while I see the usage, when I install the pkg, then look w/in that installed pkg, I do not see the actual code. Nor (which, to me, is more important), if I simply try any of the sample test codes (using PyCharm), it errors out with "module 'osxphotos' has no attribute 'PhotosDB'".

**Comment by RhetTbull** [comment]
`PhotosDB` is exported by [osxphotos.__init__](https://github.com/RhetTbull/osxphotos/blob/master/osxphotos/__init__.py) and the source is in [osxphotos/photosdb/photosdb.py](https://github.com/RhetTbull/osxphotos/blob/master/osxphotos/photosdb/photosdb.py).  My guess is something is wrong with your pycharm or python environment. Do you happen to have a folder called `osxphotos` or file called `osxphotos.py` in the folder you're working in? If you do python will look there first and that could cause the error.

**Reply by WadeJA** [reply]
Ok, thanks. Is a new Mac + a new PY installation, etc, etc. I'll look further into why this is happening. Knowing it's exported helps me better understand where & how. My apologies for such a 'NB' ? Have done lots of other dev't + a fair amt in PY; Attributing this to startup on this new box. And, no, I do not have either the folder nor the .py you reference in my setup. TY for your responses and in such quick order.

**Reply by RhetTbull** [reply]
No worries -- glad to help!  How did you install osxphotos? If doing your own dev, I recommend creating a virtual python environment, activating it, then `pip install osxphotos`.  I don't recommend cloning the GitHub repo unless you actually want to do dev on osxphotos itself. (It's a huge repo > 1GB due to test files). I use [pyenv](https://github.com/pyenv/pyenv) to manage my virtual environments but you could just use [venv](https://docs.python.org/3/tutorial/venv.html) that comes with Python. 

If you want to just run `osxphotos` as a CLI for exporting, etc. then I recommend installing with [pipx](https://github.com/pypa/pipx) so you don't have to mess with virtual environments. By the way, when installed with `pipx`, osxphotos has the ability to run arbitrary python code -- I use this all the time for prototyping osxphoto code -- and your code will have access to all the packages installed by osxphotos.  Use `osxphotos run file.py`.  You can even install packages into virtual env that pipx creates for osxphotos using `osxphotos install packagename`.

**Comment by WadeJA** [comment]
K. TY
